execute/tokendata/observer_background.go

```diff
@@ -0,0 +1,431 @@
+				msg.Header.MessageID,
+				tokenData[msg.Header.SourceChainSelector][msg.Header.SequenceNumber],
+			)
+		}
+	}
+
+}
+
+// msgQueue is a simple in-memory FIFO queue that can be used for async message processing.
+type msgQueue struct {
+	lggr             logger.Logger
+	msgs             []msgWithInfo
+	mu               *sync.RWMutex
+	newMsgSignalChan chan struct{}
+}
+
+type msgWithInfo struct {
+	msg cciptypes.Message
+	// availableAt is the time when the message can be processed
+	availableAt time.Time
+	// enqueuedAt is the time when the message was added to the queue
+	enqueuedAt time.Time
+}
+
+func newMsgQueue(lggr logger.Logger) *msgQueue {
+	return &msgQueue{
+		lggr:             lggr,
+		msgs:             make([]msgWithInfo, 0),
+		mu:               &sync.RWMutex{},
+		newMsgSignalChan: make(chan struct{}),
+	}
+}
+
+// enqueue adds a message to the queue with the given availableAfter duration.
+func (q *msgQueue) enqueue(msg cciptypes.Message, availableAfter time.Duration) bool {
+	lggr := logger.With(
+		q.lggr, "msgID", msg.Header.MessageID.String(),
+		"sourceChain", msg.Header.SourceChainSelector.String(),
+		"seqNum", msg.Header.SequenceNumber.String(),
+		"availableAfter", availableAfter,
+	)
+	lggr.Debug("waiting for the lock before adding msg")
+
+	containsMsg := q.containsMsg(msg)
+	if containsMsg {
+		lggr.Debug("message already exists in the queue")
+		return false
+	}
+
+	q.mu.Lock()
+	q.msgs = append(q.msgs, msgWithInfo{
+		msg:         msg,
+		availableAt: time.Now().Add(availableAfter).UTC(),
+		enqueuedAt:  time.Now().UTC(),
+	})
+	lggr.Debugw("message added to the queue, new msg signal sent", "numMsgs", len(q.msgs))
+	q.mu.Unlock()
+
+	lggr.Debugw("sending to new msg signal channel")
+	q.newMsgSignalChan <- struct{}{}
+	return true
+}
+
+// dequeue returns the first available message from the queue in a FIFO order.
+func (q *msgQueue) dequeue() (cciptypes.Message, bool) {
+	q.lggr.Debug("waiting for the lock before popping msg")
+
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	q.lggr.Debug("lock acquired before popping msg")
+
+	if len(q.msgs) == 0 {
+		q.lggr.Debug("no messages in the queue")
+		return cciptypes.Message{}, false
+	}
+
+	for i, msg := range q.msgs {
+		if time.Now().UTC().After(msg.availableAt) {
+			q.lggr.Debugw("message popped from the queue",
+				"msgID", msg.msg.Header.MessageID.String(),
+				"sourceChain", msg.msg.Header.SourceChainSelector.String(),
+				"seqNum", msg.msg.Header.SequenceNumber.String(),
+				"enqueuedSince", time.Since(msg.enqueuedAt),
+				"availableSince", time.Since(msg.availableAt),
+			)
+			q.msgs = append(q.msgs[:i], q.msgs[i+1:]...)
+			return msg.msg, true
+		}
+	}
+
+	return cciptypes.Message{}, false
+}
+
+// containsMsg returns true if the message is already in the queue
+func (q *msgQueue) containsMsg(msg cciptypes.Message) bool {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+
+	for _, qMsg := range q.msgs {
+		if qMsg.msg.Header.MessageID == msg.Header.MessageID {
```

Comment:
`containsMsg` has linear complexity which might become a bottleneck during a message spike scenarios. Probably not an issue right now, but may be worth adding TODO in the future. If it's easy then maybe `qMsg` should also have a set indicating message presence `map[MessageID]struct{}`